tests/PHPStan/Analyser/data/preg_match_shapes.php

```diff
@@ -0,0 +1,118 @@
+<?php
+
+namespace PregMatchShapes;
+
+use function PHPStan\Testing\assertType;
+
+
+function doMatch(string $s): void {
+	if (preg_match('/Price: (£|€)\d+/', $s, $matches)) {
+		assertType('array{string, string}', $matches);
+	} else {
+		assertType('array<string>', $matches);
+	}
+	assertType('array<string>', $matches);
+
+	if (preg_match('/Price: (£|€)(\d+)/i', $s, $matches)) {
+		assertType('array{string, string, string}', $matches);
+	}
+	assertType('array<string>', $matches);
+
+	if (preg_match('  /Price: (£|€)\d+/ i u', $s, $matches)) {
+		assertType('array{string, string}', $matches);
+	}
+	assertType('array<string>', $matches);
+
+	if (preg_match('(Price: (£|€))i', $s, $matches)) {
+		assertType('array{string, string}', $matches);
+	}
+	assertType('array<string>', $matches);
+
+	if (preg_match('_foo(.)\_i_i', $s, $matches)) {
+		assertType('array{string, string}', $matches);
+	}
+	assertType('array<string>', $matches);
+
+	if (preg_match('/(a)(b)*(c)(d)*/', $s, $matches)) {
+		assertType('array{0: string, 1: string, 2: string, 3?: string, 4?: string}', $matches);
+	}
+	assertType('array<string>', $matches);
+}
+
+function doNonCapturingGroup(string $s): void {
+	if (preg_match('/Price: (?:£|€)(\d+)/', $s, $matches)) {
+		assertType('array{string, string}', $matches);
+	}
+	assertType('array<string>', $matches);
+}
+
+function doNamedSubpattern(string $s): void {
+	if (preg_match('/\w-(?P<num>\d+)-(\w)/', $s, $matches)) {
+		// could be assertType('array{0: string, num: string, 1: string, 2: string, 3: string}', $matches);
```

Comment:
here we can see one example of a regex which is not parsable with the current hoa grammar:

```
Fatal error: Uncaught Hoa\Compiler\Llk\Parser::parse(): (0) Unexpected token "-" (range) at line 1 and column 4:
/\w-(?P<num>\d+)-(\w)/
   ↑
in /Users/staabm/workspace/phpstan-src/vendor/hoa/compiler/Llk/Parser.php at line 1.
  thrown in /Users/staabm/workspace/phpstan-src/vendor/hoa/compiler/Llk/Parser.php on line 1
```

Reply:
There are two issues:

a) `(?P<xxx>...` is equivalent of supported `(?<xxx>...` - spec https://www.pcre.org/original/doc/html/pcrepattern.html (search for ˙?p<˙), grammar: https://github.com/hoaproject/Regex/blob/master/Source/Grammar.pp#L73
b) `-` range - it is reserved character strictly if in `[` character group only (otherwise it is regular chanracter and very often used)

Reply:
Thanks. Please provide code suggestions for fixes.

Reply:
Would you mind forking `hoaproject/Regex` and saving there a php test file I can run with all the issues you found?

Reply:
looking into this PR here and find all commented assertions. these are the cases which hoa cannot parse.

I have spent the last the 3 days of freetime for your feature request. take your time to do the tests yourself.

Reply:
fix is here: https://github.com/mvorisek/Hoa-Regex/commit/0c8a5cbcf696df8f65168704af14ad5ebd48c641

Reply:
I tried your suggestion, but it did not fix any of the commented assertions.

parsing `'/\w-(?P<num>\d+)-(\w)/'` I still get an exception:
```
PHP Fatal error:  Uncaught Hoa\Compiler\Llk\Parser::parse(): (0) Unexpected token "?" (zero_or_one) at line 1 and column 6:
/\w-(?P<num>\d+)-(\w)/
     ↑
in /Users/staabm/workspace/phpstan-src/vendor/hoa/compiler/Llk/Parser.php at line 1.
  thrown in /Users/staabm/workspace/phpstan-src/vendor/hoa/compiler/Llk/Parser.php on line 1

```

Reply:
It is because of the `(?P<xxx>...)` part as explained above - the `-` in non-class is fixed. It should be present as a separate case.

Reply:
ohh I see. I got the impression you are working on all these cases.

Reply:
I can look into more cases later. This one was the most important. What about adding proper Hoa Regex unit testing to phpstan-src? We can assert the parsed regexes using https://github.com/mvorisek/Hoa-Regex/blob/0c8a5cbcf696df8f65168704af14ad5ebd48c641/test.php#L13-L25 function.

Reply:
Sure, why not. Please send a PR to my branch

Reply:
I am not super familiar with phpstan-src codebase - where, what class can I base this unit test on?

Reply:
I would create a new file in `tests/PHPStan/Type`